parser/parser.go

```diff
@@ -0,0 +1,252 @@
+		}
+		var data []bson.M
+		match :=
+			bson.D{
+				{"$match",
+					bson.D{
+						{"course_number", classes[i].CatalogNumber},
+						{"subject_prefix", classes[i].Subject},
+					},
+				},
+			}
+		lookup :=
+			bson.D{
+				{"$lookup",
+					bson.D{
+						{"from", "sections"},
+						{"localField", "sections"},
+						{"foreignField", "_id"},
+						{"as", "sections"},
+					},
+				},
+			}
+		unwind := bson.D{{"$unwind", bson.D{{"path", "$sections"}}}}
+		matchSectionNo :=
+			bson.D{
+				{"$match",
+					bson.D{
+						{"sections.section_number", classes[i].Section},
+						{"sections.academic_session.name", academicSession},
+					},
+				},
+			}
+		project := bson.D{{"$project", bson.D{{"section", "$sections"}}}}
+		set :=
+			bson.D{
+				{"$set",
+					bson.D{
+						{"section.grade_distribution", classes[i].GradeDistribution},
+					},
+				},
+			}
+		cursor, err := coursesCollection.Aggregate(context.TODO(), mongo.Pipeline{match, lookup, unwind, matchSectionNo, project, set})
+		if err != nil {
+			logFile.WriteString(classes[i].Subject + " " + classes[i].CatalogNumber + "." + classes[i].Section + ": " + err.Error() + "\n")
+			fmt.Println(classes[i].Subject + " " + classes[i].CatalogNumber + "." + classes[i].Section + ": " + err.Error())
+			continue
+		}
+		err = cursor.All(context.TODO(), &data) // put cursor data into []primitive.M
+		if err != nil {
+			logFile.WriteString(classes[i].Subject + " " + classes[i].CatalogNumber + "." + classes[i].Section + ": " + err.Error() + "\n")
+			fmt.Println(classes[i].Subject + " " + classes[i].CatalogNumber + "." + classes[i].Section + ": " + err.Error())
+			continue
+		}
+		// if more than 1 result, log error and continue
+		if len(data) != 1 {
+			logFile.WriteString(classes[i].Subject + " " + classes[i].CatalogNumber + "." + classes[i].Section + ": recieved results " + strconv.Itoa(len(data)) + " from aggregation, expected 1\n")
+			fmt.Println(classes[i].Subject + " " + classes[i].CatalogNumber + "." + classes[i].Section + ": recieved results " + strconv.Itoa(len(data)) + " from aggregation, expected 1")
+			continue
+		}
+
+		section, ok := data[0]["section"].(primitive.M)
+		if !ok { // if section assertion not work
+			logFile.WriteString(classes[i].Subject + " " + classes[i].CatalogNumber + "." + classes[i].Section + "unable to get section from db")
+			fmt.Println(classes[i].Subject + " " + classes[i].CatalogNumber + "." + classes[i].Section + "unable to get section from db")
+			continue
+		}
+		sectionID, ok := section["_id"].(primitive.ObjectID)
+		if !ok { // if sections id assertion not work
+			logFile.WriteString(classes[i].Subject + " " + classes[i].CatalogNumber + "." + classes[i].Section + "unable to get section id from db")
+			fmt.Println(classes[i].Subject + " " + classes[i].CatalogNumber + "." + classes[i].Section + "unable to get section id from db")
+		}
+		var gradeDistribution bson.A = data[0]["section"].(primitive.M)["grade_distribution"].(bson.A)
```

Comment:
Did you mean to put `section["grade_distribution"]` here? Since you've already defined it before? Also you should check this assertion as well to avoid a panic.